### PR TITLE
Fix: Route liger student forward through DDP wrapper in GKD, GOLD, and Distillation trainers

### LINKED ISSUE #5282: [Question] Do use_liger_gkd_loss breaks DDP gradient sync by unwrapping student model？
In the GKD and GOLD trainers, when `use_liger_gkd_loss` is enabled, the code explicitly unwraps the student model before the forward pass:

```python
unwrapped_student = self.accelerator.unwrap_model(model)
if hasattr(unwrapped_student, "get_decoder") and unwrapped_student.get_decoder() is not None:
    base_student = unwrapped_student.get_decoder()
else:
    base_student = getattr(
        unwrapped_student, getattr(unwrapped_student, "base_model_prefix", "model"), unwrapped_student
    )

student_outputs = base_student(
    input_ids=inputs["input_ids"],
    attention_mask=inputs["attention_mask"],
    use_cache=False,
)
```
Isn't this problematic for distributed training? By bypassing the DistributedDataParallel wrapper for the student's forward pass, the gradients won't be synchronized across GPUs via AllReduce, causing the model weights on different ranks to diverge. Shouldn't the student model be called directly (e.g., model(...)) to preserve the DDP graph?

### PR DESCRIPTION
**Pull request by albertvillanova**
Fix: Route liger student forward through DDP wrapper in GKD, GOLD, and Distillation trainers.

This PR fixes the Liger JSD loss computation in the distillation, GKD, and GOLD trainers to improve memory efficiency and ensure compatibility with DDP distributed training setups. The main changes involve introducing a `_ForwardRedirection` utility to correctly route forward passes through distributed wrappers, and consolidating the decoder-only forward logic into a helper method. These changes help avoid unnecessary computations, reduce memory usage, and ensure gradients are handled correctly.

Fix #5282.

### Motivation

When `use_liger_kernel=True`, the GKD, GOLD, and Distillation experimental trainers extracted the decoder sub-module and called it directly on the unwrapped model:
```python
unwrapped_student = self.accelerator.unwrap_model(model)
base_student = unwrapped_student.get_decoder()
student_outputs = base_student(...)  # bypasses DDP.forward()
```

This skips `DDP.forward()`, so `Reducer.prepare_for_backward()` is never called. Under DDP this breaks per-iteration gradient accounting, with `find_unused_parameters=True`, it silently misroutes AllReduce; without it, PyTorch raises Expected to have finished reduction in the prior iteration.

The SDFT and SDPO trainers already handled this correctly via `_ForwardRedirection`, which patches the unwrapped module's `forward` and then calls the DDP wrapper so that `prepare_for_backward()` fires, while still executing only the decoder (no lm_head) inside. This PR applies the same pattern to the three affected trainers.

### Changes

**Distributed training and forward redirection improvements:**

* Introduced the `_ForwardRedirection` utility in `distillation_trainer.py`, `gkd_trainer.py`, and `gold_trainer.py` to ensure that forward passes for Liger JSD loss are routed through DDP/FSDP wrappers, enabling correct gradient handling and memory efficiency.
* Instantiated `_ForwardRedirection` in the constructors of all three trainers for use during Liger JSD loss computation. 

**Memory-efficient decoder-only forward logic:**

* Refactored the decoder-only forward logic into a new `_liger_student_forward` helper method in each trainer, allowing forward passes that skip the `lm_head` to save memory during Liger JSD loss computation. 
* Updated the Liger JSD loss computation in each trainer to use `_ForwardRedirection` and `_liger_student_forward`, replacing the previous manual handling of decoder extraction and forward calls. 

These changes improve both the efficiency and correctness of the Liger JSD loss path in distributed and memory-constrained training scenarios.

---

> [!NOTE]
> **Medium Risk**
> Changes distributed training backward hooks for the Liger path only; wrong routing could still break DDP, but the fix aligns with an existing pattern elsewhere in TRL.
> 
> **Overview**
> When **`use_liger_kernel=True`**, the Distillation, GKD, and GOLD trainers no longer call the unwrapped student **decoder** directly for Liger JSD. They now route that forward through **`_ForwardRedirection`** so **DDP/FSDP** `forward()` runs and **`prepare_for_backward()`** is registered—matching SDFT/SDPO and fixing broken or incorrect multi-GPU training.
> 
> Each trainer creates **`_ForwardRedirection()`** at init when Liger is enabled and adds **`_liger_student_forward`** for decoder-only passes (still skipping **`lm_head`** for memory). **`_compute_liger_loss`** / the Liger branch in **`compute_loss`** invokes **`self._forward_redirection(model, unwrapped_student, self._liger_student_forward, ...)`** instead of **`base_student(...)`** on the unwrapped module.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3c1d658dae61e5a0bf5de86cedf9a06767991524. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
awesome thanks checking!

**Comment by kashif** [comment]
@albertvillanova i have a largish gold trainer refactor: https://github.com/huggingface/trl/pull/5885 so do you think i will get a merge conflict if we go with the fix for GOLD?

**Comment by kashif** [comment]
I checked and should be fine!

**Review by kashif — APPROVED** [review]
